carddav/carddav_test.go

```diff
@@ -0,0 +1,173 @@
+END:VCARD`
+	alicePath = "urn:uuid:4fbe8971-0bc3-424c-9c26-36c3e1eff6b1.vcf"
+
+	currentUserPrincipalKey = "test:currentUserPrincipal"
+	homeSetPathKey          = "test:homeSetPath"
+	addressBookPathKey      = "test:addressBookPath"
+)
+
+func (*testBackend) CurrentUserPrincipal(ctx context.Context) (string, error) {
+	r, ok := ctx.Value(currentUserPrincipalKey).(string)
+	if !ok {
+		return "", fmt.Errorf("missing currentUserPrincipal in context")
+	}
+	return r, nil
+}
+
+func (*testBackend) AddressbookHomeSetPath(ctx context.Context) (string, error) {
+	r, ok := ctx.Value(homeSetPathKey).(string)
+	if !ok {
+		return "", fmt.Errorf("missing homeSetPath in context")
+	}
+	return r, nil
+}
+
+func (*testBackend) AddressBook(ctx context.Context) (*AddressBook, error) {
+	p, ok := ctx.Value(addressBookPathKey).(string)
+	if !ok {
+		return nil, fmt.Errorf("missing addressBookPath in context")
+	}
+	return &AddressBook{
+		Path:                 p,
+		Name:                 "My contacts",
+		Description:          "Default address book",
+		MaxResourceSize:      1024,
+		SupportedAddressData: nil,
+	}, nil
+}
+
+func (*testBackend) GetAddressObject(ctx context.Context, path string, req *AddressDataRequest) (*AddressObject, error) {
+	if path == alicePath {
+		card, err := vcard.NewDecoder(strings.NewReader(aliceData)).Decode()
+		if err != nil {
+			return nil, err
+		}
+		return &AddressObject{
+			Path: path,
+			Card: card,
+		}, nil
+	} else {
+		return nil, webdav.NewHTTPError(404, fmt.Errorf("Not found"))
+	}
+}
+
+func (b *testBackend) ListAddressObjects(ctx context.Context, req *AddressDataRequest) ([]AddressObject, error) {
+	alice, err := b.GetAddressObject(ctx, alicePath, req)
+	if err != nil {
+		return nil, err
+	}
+
+	return []AddressObject{*alice}, nil
+}
+
+func (*testBackend) QueryAddressObjects(ctx context.Context, query *AddressBookQuery) ([]AddressObject, error) {
+	panic("TODO: implement")
+}
+
+func (*testBackend) PutAddressObject(ctx context.Context, path string, card vcard.Card, opts *PutAddressObjectOptions) (loc string, err error) {
+	panic("TODO: implement")
+}
+
+func (*testBackend) DeleteAddressObject(ctx context.Context, path string) error {
+	panic("TODO: implement")
+}
+
+func TestAddressBookDiscovery(t *testing.T) {
+
+	for _, tc := range []struct {
+		name                 string
+		currentUserPrincipal string
+		homeSetPath          string
+		addressBookPath      string
+	}{
+		// TODO this used to work, but is currently broken.
+		//{
+		//	name:  "all-at-root",
+		//	currentUserPrincipal: "/",
+		//	homeSetPath: "/",
+		//	addressBookPath: "/",
+		//},
+		{
+			name:                 "simple-home-set-path",
+			currentUserPrincipal: "/",
+			homeSetPath:          "/contacts/",
+			addressBookPath:      "/contacts/",
+		},
+		{
+			name:                 "explicit-well-known",
+			currentUserPrincipal: "/test/",
+			homeSetPath:          "/test/contacts/",
+			addressBookPath:      "/test/contacts/",
+		},
```

Comment:
Maybe we can have another test case with separate URIs for home-set and addressbook?

```go
		{
			name:                 "all-separate-paths",
			currentUserPrincipal: "/test",
			homeSetPath:          "/test/contacts",
			addressBookPath:      "/test/contacts/my-address-book",
		},
```

Reply:
added